### PR TITLE
Enforce Check-Effects-Interaction throughout EMP

### PR DESCRIPTION
**Pull request by nicholaspai**
Some methods still have produce effects post-interaction.

`dispute` and `payFees` performed the most effects post-interaction but they were also lower-severity reentrancy vulnerabilities given that they do not send tokens out of the contract.

### CODE REVIEW

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 1]
```diff
@@ -47,6 +47,8 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
         // Raw collateral value. This value should never be accessed directly -- always use _getCollateral().
         // To add or remove collateral, use _addCollateral() and _removeCollateral().
         FixedPoint.Unsigned rawCollateral;
+        // Tracks pending transfer requests. A transfer request is pending if `transferRequestPastTimestamp != 0`.
+        uint256 transferRequestPassTimestamp;
```
Comment (nicholaspai) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol:51]: I unintentionally forked this branch off of the branch I used in #1314 so once that PR is merged this file will show a much smaller diff

Comment (nicholaspai) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: done

**Comment by mrice32** [comment]
@nicholaspai have we made a decision on whether to use OZ's re-entrancy checker?

**Comment by nicholaspai** [comment]
> @nicholaspai have we made a decision on whether to use OZ's re-entrancy checker?

We have not finalized a decision, but after discussing with OZ, I recommend we don't add it in this PR for several reasons:
- Simplicity: fewer functions/modifiers to worry about, lower gas cost without the modifier. 
- Interactions between functions with the reentrancy guard modifier: If we add this modifier to `payFees`, then because every function also calls `payFees` via the `fees()` modifier, it is possible the contract interactions will break. According to OZ: "functions marked as `nonReentrant` may not call one another. This can be worked around by making those functions `private`, and then adding `external` `nonReentrant` entry points to them."
- Security: `dispute` and `payFees` do not remove tokens from the contract so they are lower severity than `withdrawPassedRequest` for example.

However, I think adding the reentrancy guard throughout the repo is a good idea but we'd need to do a deeper dive into it. 

I wanted to just make sure that we were at least following the CEI pattern in this PR.

**Comment by mrice32** [comment]
> > @nicholaspai have we made a decision on whether to use OZ's re-entrancy checker?
> 
> We have not finalized a decision, but after discussing with OZ, I recommend we don't add it in this PR for several reasons:
> 
> * Simplicity: fewer functions/modifiers to worry about, lower gas cost without the modifier.
> * Interactions between functions with the reentrancy guard modifier: If we add this modifier to `payFees`, then because every function also calls `payFees` via the `fees()` modifier, it is possible the contract interactions will break. According to OZ: "functions marked as `nonReentrant` may not call one another. This can be worked around by making those functions `private`, and then adding `external` `nonReentrant` entry points to them."
> * Security: `dispute` and `payFees` do not remove tokens from the contract so they are lower severity than `withdrawPassedRequest` for example.
> 
> However, I think adding the reentrancy guard throughout the repo is a good idea but we'd need to do a deeper dive into it.
> 
> I wanted to just make sure that we were at least following the CEI pattern in this PR.

Seems reasonable. I just don't really trust the CEI pattern to protect us against re-entrancy (see the Uniswap V1 imBTC example). If our contract is going to be used with tokens that can re-enter, I think we should either add re-entrancy guards or document that it isn't compatible with ERC777 or ERC223.

**Comment by nicholaspai** [comment]
> > > @nicholaspai have we made a decision on whether to use OZ's re-entrancy checker?
> > 
> > 
> > We have not finalized a decision, but after discussing with OZ, I recommend we don't add it in this PR for several reasons:
> > 
> > * Simplicity: fewer functions/modifiers to worry about, lower gas cost without the modifier.
> > * Interactions between functions with the reentrancy guard modifier: If we add this modifier to `payFees`, then because every function also calls `payFees` via the `fees()` modifier, it is possible the contract interactions will break. According to OZ: "functions marked as `nonReentrant` may not call one another. This can be worked around by making those functions `private`, and then adding `external` `nonReentrant` entry points to them."
> > * Security: `dispute` and `payFees` do not remove tokens from the contract so they are lower severity than `withdrawPassedRequest` for example.
> > 
> > However, I think adding the reentrancy guard throughout the repo is a good idea but we'd need to do a deeper dive into it.
> > I wanted to just make sure that we were at least following the CEI pattern in this PR.
> 
> Seems reasonable. I just don't really trust the CEI pattern to protect us against re-entrancy (see the Uniswap V1 imBTC example). If our contract is going to be used with tokens that can re-enter, I think we should either add re-entrancy guards or document that it isn't compatible with ERC777 or ERC223.

I'm going to try to add them in a follow up PR, relevant to this [issue](https://github.com/UMAprotocol/protocol/issues/1332)

**Pull request merged**